### ISSUE TITLE
CrtSwitchres not working 1.9.1

### ISSUE THREAD

**Issue by Redemp** [libretro/RetroArch#12244]
## Description

[Description of the bug]
CrtSwitchres not working

### Expected behavior
CrtSwitchres should switch resolution

[What you expected to happen]

CrtSwitchres should switch resolution

### Actual behavior
[What is actually happening]
CrtSwitchres not working in anything except native resolution

### Bisect Results

- RetroArch: [version/commit]
1.9.1
### Environment information
Linux

**Comment by twinaphex** [comment]
hi there @alphanu1 , could you possibly take a look to see if there is an actual problem going on and a regression happened somewhere?

**Comment by alphanu1** [comment]
@twinaphex  I have just span up Linus testing environment. I can confirm that CRTSwitchRes is working on the latest build. All options are functioning as they should, allowing all types of super resolutions (Native, Dynamic, 1920, 2560 and 3840).

**Enviroment**
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

**List of resolutions tested with CRTSwitchRes**
```
 CRT1 (0x4ef)  6.139MHz -HSync -VSync
        h: width   320 start  330 end  357 total  392 skew    0 clock  15.66KHz
        v: height  240 start  242 end  245 total  261           clock  60.00Hz
  CRT2 (0x4f1) 31.116MHz -HSync -VSync
        h: width  1600 start 1666 end 1787 total 1987 skew    0 clock  15.66KHz
        v: height  240 start  242 end  245 total  261           clock  60.00Hz
  CRT3 (0x4f2) 37.349MHz -HSync -VSync
        h: width  1920 start 2000 end 2144 total 2385 skew    0 clock  15.66KHz
        v: height  240 start  242 end  245 total  261           clock  60.00Hz
  CRT4 (0x4f5) 49.799MHz -HSync -VSync
        h: width  2560 start 2666 end 2859 total 3180 skew    0 clock  15.66KHz
        v: height  240 start  242 end  245 total  261           clock  60.00Hz
  CRT5 (0x4f6) 74.698MHz -HSync -VSync
        h: width  3840 start 4000 end 4289 total 4770 skew    0 clock  15.66KHz
        v: height  240 start  242 end  245 total  261           clock  60.00Hz
```

I have seen some comments of the forum about this too. More investigation is needed to confirm why some users are having issues.

@Redemp Please can you give me some more information.

What Linux distribution are you using?
Was it working on this same install before?
What graphics card are you using, did it work before?
Have you deleted your retroarch.cfg so a new one can be generated after upgrading, including core overrides?

**Comment by Redemp** [comment]
Reinstalled everything and now it works. I think i was using outdated cores but i have no idea. So strange... :man_facepalming:

Wrote on the forum so hopefully someone else reports back so this could be an issue with old installation files aka user error. 
Sorry to cry wolf for nothing.

**Comment by alexb3d** [comment]
Currently it does not work for me in any way, sometimes close the program other times the screen goes black and the only way is to force restart.
I have deleted the configuration folder and installed the kernels again, the error continues.

Testing with the PPA (stable) and with Flatpak. 
I use Ubuntu 20.04.2, Core i7 3770 and the integrated HD Graphics 4000

It was working perfect for me with a Monitor, just by activating 31Hz it created all the resolutions I needed, 240p 480i, etc ...

This I wrote 20 days ago in[ the forum](https://forums.libretro.com/t/new-crtswitchres-update/30087/10).

**Comment by alphanu1** [comment]
Discussions ongoing on forum. [https://forums.libretro.com/t/new-crtswitchres-update/30087/46](url)

> @alexb3d  It looks like the resolution has been created and added to an out put. so it seems something is happening just after that process has happened. You would not see CRT1 if it crashed while creating the resolution.
> 
> Can you run RA in verbose mode please. Output to a log file and attach it here.
> 
> retroarch --menu --verbose >> retroarch.log 2>&1

**Comment by Kopert** [comment]
I am having the same problem on a Linux (Manjaro) machine where I have been running RetroArch with switchres for the past year with no problems. I just got 1.9.1 today.

> Can you run RA in verbose mode please. Output to a log file and attach it here.
> retroarch --menu --verbose >> retroarch.log 2>&1

Sure, [retroarch.log](https://github.com/libretro/RetroArch/files/6286507/retroarch.log)

Relevant part seems to be:

`X Error of failed request:  BadAccess (attempt to access private resource denied)`
`Major opcode of failed request:  140 (RANDR)`
`Minor opcode of failed request:  19 (RRDeleteOutputMode)`
`Serial number of failed request:  112`
`Current serial number in output stream:  114`

**Comment by alphanu1** [comment]
@Kopert I have not tested Manjaro yet. But if it worked in the past it should be fine now.

I am in the process of download a Manjaro ISO. However, source forge is running unbelievably slow. 10+ hours ATM.

@alexb3d are you using Ubuntu?

**Comment by Kopert** [comment]
> @Kopert I have not tested Manjaro yet. But if it worked in the past it should be fine now.
> 
> I am in the process of download a Manjaro ISO. However, source forge is running unbelievably slow. 10+ hours ATM.

Thank you for taking the time to look into this. If it is of any relevance, I'm using kernel 5.4 (LTS) on Manjaro because the newer kernels (5.10 onward) do not support VGA outputs on AMD cards. It's the same kernel I've used for the past year.

Manjaro has always been good to me for running Retroarch. I had more luck with it than with other distros.

Let me know if there is any way I can help you investigate this.

**Comment by Kopert** [comment]
Additional info: just to isolate 1.9.1 as the main problem, I located the 1.9.0 package (found it in my local pacman cache as `retroarch-1.9.0-2-x86_64.pkg.tar.zst`), extracted the old `retroarch` binary and executed it directly, and everything is working as it used to. Re-running 1.9.1 still crashes on startup as before.

**Comment by alphanu1** [comment]
I will be compiling fresh once I have the Manjaro Environment setup. Please see tests below. I will update this post as I go along.

Tests:
- Ubuntu 20.04 RA 1.9.1 running CRTSwichRes [https://youtu.be/0aRragPH1I0](url)

**Comment by Kopert** [comment]
> You have to compile your own.
> Look here https://github.com/D0023R/linux_kernel_15khz

Oh no, it works on pre-5.10 kernels. The problem is that they added DC support for my GPU on 5.10 but it still does not support VGA outputs as you can see [here](https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.10-Feature-Recap) and [here](https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.10-Graphics-Drivers).

I'm staying on 5.4 LTS and it works fine. I'll consider upgrading if/when they manage to get VGA output working.

**Comment by alphanu1** [comment]
@Kopert are you using a patch kernel then?

**Comment by Kopert** [comment]
> @Kopert are you using a patch kernel then?

No, never used a patch kernel, I'm using everything vanilla on a 5.4 kernel. 1.9.0 still works.

**Comment by Redemp** [comment]
> > You have to compile your own.
> > Look here https://github.com/D0023R/linux_kernel_15khz
> 
> Oh no, it works on pre-5.10 kernels. The problem is that they added DC support for my GPU on 5.10 but it still does not support VGA outputs as you can see [here](https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.10-Feature-Recap) and [here](https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.10-Graphics-Drivers).
> 
> I'm staying on 5.4 LTS and it works fine. I'll consider upgrading if/when they manage to get VGA output working.

Just disabled DC in kernel command. What type of card do you have? DVI-I port? The go DVI-I to VGA.

**Comment by Kopert** [comment]
> Just disable DC in kernel command during boot

I lack the ability to mess with that sort of stuff. Last time I tried, everything got broken and I had to reinstall Linux from scratch. Also, I can get Vulkan to work with my GPU pre-5.10; I'm pretty sure it won't work on 5.10 if I disable DC. But it doesn't matter, 5.4 is supported until December 2025 and is marked as "recommended" in Manjaro. There is no reason for me to change to a different kernel at this point.

> What type of card do you have? DVI-I port? The go DVI-I to VGA.

The DVI port on my GPU does not have the analog pins, so going DVI->VGA is not possible.

Anyway, I think this is off-topic to the current RetroArch issue. It has always worked on kernel 5.4.

**Comment by Kopert** [comment]
So I decided to try running kernel 5.10 with DC disabled, and there is no difference in behavior. 1.9.0 works, 1.9.1 doesn't.

I have noticed something that occurred the first time but I hadn't realized before. The **first** time RetroArch 1.9.1 is run, it freezes during startup with garbage on the screen after setting a super resolution (I use 2560). I can't close it unless I do a kill -9. The error messages from the first log I sent are from trying to run RetroArch again after this first failure.

In case this is relevant, I restarted my computer on kernel 5.4 and ran RetroArch to grab a log from the first time it runs (when it freezes on a garbled screen), [here it is](https://github.com/libretro/RetroArch/files/6287373/retroarch.log). I have to `kill -9` it to get it to close.

**Comment by Kopert** [comment]
I tried switching to "gl" for the video driver (instead of vulkan) and I got image only on a vertical strip on the left side; the rest is a garbled mess. Trying to start a game causes a crash with the error message from my first log. Trying to run 1.9.0 after this still works fine.
![IMG_20210409_141652](https://user-images.githubusercontent.com/26443632/114217234-4768ba80-993e-11eb-92f2-21bdb8cb4dd6.jpg)

**Comment by alphanu1** [comment]
> I will be compiling fresh once I have the Manjaro Environment setup. Please see tests below. I will update this post as I go along.
> 
> Tests:
> 
> * Ubuntu 20.04 RA 1.9.1 running CRTSwichRes [https://youtu.be/0aRragPH1I0](url)
> * Manjaro 20.2.1 RA 1.9.1 runiing CRTSwitchRes [https://youtu.be/2dWBGgY1Dvo](url)

@Kopert

**Comment by Kopert** [comment]
@alphanu1 I understand, but I'm still getting the problem here. I assume you saw my previous messages with the new log for the first-run freeze.

I tried downloading a brand new AppImage version of RetroArch 1.9.1 with its own fresh settings. Changed the menu driver to rgui, enabled advanced settings and restarted RetroArch. But as soon as I enable switchres all I get is a vertical strip of working video on the left and shimmering garbage on the rest of the screen. If I try to start a game, it crashes; if I switch to Vulkan, it crashes as soon as I turn switchres on.

For now I think we can rule out kernel version (I tried 5.4 and 5.10) and distro (since you can run it on Manjaro). We can also rule out other RetroArch settings/data since I got a fresh AppImage version with fresh settings.

Right now I'm not sure what else I can try to workaround the problem other than just stick to 1.9.0. If you have any other ideas of stuff for me to try I'd appreciate it.

**Comment by alphanu1** [comment]
@Kopert  can you try and compile RA instead of using the APP image?

**Comment by Kopert** [comment]
@alphanu1 I don't think I can do it but I'll try.

**Comment by alphanu1** [comment]
Open upo a terminal window and follow these commands.
```
sudo pacman -S base-devel
git clone https://github.com/libretro/RetroArch
cd RetroArch
./configure
make
./retroarch 
```
While testing every time you want to load RetoArch run ./retroarch within the compiled folder.

**Comment by Kopert** [comment]
@alphanu1 I followed your instructions. Unfortunately, the behavior is the same - it crashes if I try to use Vulkan, and using OpenGL it only shows the left part of the image, with the rest of the TV flashing some garbled mess.

On the attached screenshot you can see that RetroArch is being cut-off. The TV is to the right of the HDMI monitor.

![Screenshot_20210409_162941](https://user-images.githubusercontent.com/26443632/114231647-147bf200-9951-11eb-9a0c-ce602843665b.png)

**Comment by Kopert** [comment]
@alphanu1 I'm not sure this is useful, but I was trying to pinpoint the location of the issue.

If I take the git repository on the most recent version and I do

`git checkout daf6843 -- gfx/display_servers/dispserv_x11.c && make && ./retroarch`

It works. If I do

`git checkout 877408a -- gfx/display_servers/dispserv_x11.c && make && ./retroarch`

It crashes on startup. If I do

`git checkout b33f0b9 -- gfx/display_servers/dispserv_x11.c && make && ./retroarch`

It does not crash on startup, but it does the "vertical-strip-with-garbled-mess" and crashes when starting a game.

**Comment by jonnyapps** [comment]
> @alphanu1 I followed your instructions. Unfortunately, the behavior is the same - it crashes if I try to use Vulkan, and using OpenGL it only shows the left part of the image, with the rest of the TV flashing some garbled mess (trying to start a game causes a crash).
> 
> On the attached screenshot you can see that RetroArch is being cut-off. The TV is to the right of the HDMI monitor.
> 
> ![Screenshot_20210409_162941](https://user-images.githubusercontent.com/26443632/114231647-147bf200-9951-11eb-9a0c-ce602843665b.png)
> 
> Edit: I tried moving the TV to the left of the monitor; vulkan still crashes (and starting a game still crashes) but RetroArch is shown fully on the TV - however, part of it is being replicated on the main monitor as well.
> 
> Is there a chance that the new switchres code is trying to manage the resolution on my main monitor alongside my TV, and is failing in doing so? Or trying to use both as a single screen? Just spitballing here.
> 
> This is how it looks when the TV is changed to be on the left side of the main monitor:
> 
> ![Screenshot_20210409_163723](https://user-images.githubusercontent.com/26443632/114232306-0d091880-9952-11eb-81fc-913f2bb02b8e.png)
> 
> Edit 2: the behavior on this second screenshot also happens on 1.9.0 (if set to the left of the main screen, retroarch extends into it) but it does not prevent anything else from working properly.
> 
> Edit 3: disabling the monitor and leaving only the TV connected does not fix the issue. Disabling the main monitor completely, restarting the computer and going into RetroArch in OpenGL works (no cut-off), but crashes as soon as I try to start a game.

I'm having the exact same issue as described here. The only difference is I have no idea what I'm doing when it comes to Linux, whereas @kopert does! I do know at least that I'm on Ubuntu 20.10

**Comment by Kopert** [comment]
I was looking at dispserv_x11.c and I found some references to "HDMI-0", "DVI-0" and such. On my system, however, xrandr identifies my ouputs as "HDMI-A-0", "DVI-D-0" and "VGA-0", in this order, with DVI-D-0 disconnected. I have not changed this in any way nor do I know how to go about changing it. But maybe this difference in naming has something to do with the issue?

It seems I figured out that the crash occurs because the code between lines 616/638 of dispserv_x11.c tries to delete the mode from HDMI-A-0 first. If, instead of letting the code iterate over `res->noutput` I just set `j` to `2` (pointing it to my VGA-0 output) it deletes the mode correctly without crashing.

I haven't yet figured out, however, why the resolution is cut-off horizontally. I'll keep poking at the code but I don't know where exactly to look yet.

**Comment by alphanu1** [comment]
@Kopert That code has not change between 1.9.0 and 1.9.1. As a side not to this. No resolution are deleted until you exit RetroArch. this is done as a clean up process. Also these resolution are only temporarily added to the outputs and do not remain after a system restart.

@twinaphex 
There are no code changes for CRTSwitchRes between 1.9.0 and 1.9.1. So, this potentially could be a video driver change causing the crash on enabling CRTSwitchRes!?

I have never tested Vulkan and have only ever used OpenGL or DirectX video drivers. So, I would announce that Vulkan is not supported even if it does work in some setups.

I am unable to replicate any of these issues mentioned above on Ubuntu 18 or 20, Lubuntu 18 or 19, Arch or Manjaro. From fresh installs to existing installs. Cloning the latest git into a new folder, compiling and running CRTSwichRes always works.

**Note (ToDo List): There is no regression. However, these will be nice to haves.**
```
Desktop restore resolution only restore to a 15kz mode. This will leave a black screen on 31khz mode.
On cleanup resolutions. Check mode exists on output before trying to remove it. (This is only a problem in a multi monitor setup. CRTSwitchRes was never designed to fully function in a multi monitor environment.)
```

**Comment by Kopert** [comment]
@alphanu1 the mode deletion code **does** run while RetroArch is running, on resolution changes, that's when it crashes. Deleting those two lines prevents the crashes, changing the code to not iterate the outputs and delete the resolutions directly from my VGA-0 also prevents them. But it **does** run during normal usage, before exiting RetroArch, otherwise it wouldn't crash during execution or when starting a game.

Regarding the cut-off, I've found that setting the VGA-0 resolution to 2560x480 using xrandr **before** starting RetroArch solves it. I think that there is a difference between what the xrandr command does (as the old switchres functioned) and the library method the new switches uses.

Also, the new switchres code is causing a lot more flicker than before when switching resolutions. Vulkan is indeed unusable on 1.9.1, but I can understand if you consider it unsupported.

Finally, I have no intention on disagreeing with a contributor that develops code that I've used so much over the past year, and I do not want to sound confrontational. However, I think there **are** regressions. Reverting the `dispserv_x11.c` to the previous version allows me to use 1.9.1 the same way I used 1.9.0, with no crashes, no horizontal cut-offs, less flicker, Vulkan support and no problems with multi-monitor setups. I completely understand if you prefer to move forward with this new code, but, for people currently having problems, the old one behaves a lot better.

Of course, since you can't replicate it and it doesn't seem to be a widespread problem then I'll have to settle with compiling the older switchres code in and hoping it doesn't break compatibility in the future.

Thanks for your time and sorry if I couldn't be of more help.

**Comment by Kopert** [comment]
Specifically regarding the horizontal cut-off, I figured that the reported "Screen 0" size by xrandr is set to 2560x480 if I set to a mode of my own that is 2560 wide; the mode set by RetroArch is 700x480, which causes the cut-off. This is with only the VGA-0 output connected, it isn't a multi-monitor issue. The resolution on the TV is correct and the window size of RetroArch is correct, it's just the screen size that is wrong.

If I start RetroArch after setting my mode that is 2560 wide, it works fine, even when restarting RetroArch.

![Screenshot_20210411_161820](https://user-images.githubusercontent.com/26443632/114318097-ccd8a000-9ae1-11eb-8d9c-13c486c2ce7d.png)

**Comment by alphanu1** [comment]
@Kopert Sorry you are correct. The clean does happen when RA is running (kind of) RA frontend has to exit to load the core, cleanup will happen at the point.

I have had `dispserv_x11.c` from 1.9.0 and 1.9.1. There is no change in the create or delete resolutions code. The file has not been changed for 7 months.

Code does need to be changed/implemented to fully support multimonitor. especially with the deletion of modes. 

I am not sure what you mean about the horizontal issue? Have you ensured you have enabled core provided aspect ratio?

**Comment by Kopert** [comment]
> I have had `dispserv_x11.c` from 1.9.0 and 1.9.1. There is no change in the create or delete resolutions code. The file has not been changed for 7 months.

The switchres code in that file has changed between 1.9.0 and 1.9.1, no? While the file has not been changed since _September 2020_, 1.9.0 - the version that we're comparing to, where eveything worked - is from _August 2020_. The file has definitely changed between 1.9.0 and 1.9.1. The code to create and delete resolutions has, to my understanding, changed in 1.9.1, as has all the switchres code in that file.

> I am not sure what you mean about the horizontal issue? Have you ensured you have enabled core provided aspect ratio?

This is a problem that happens as soon as you start RetroArch, with no cores involved, but yes, that option is enabled, but unrelated. I'll try to explain it as I understand it right now:

When I start RetroArch, the **display resolution** (the TV connected at VGA-0) is set correctly to a width of 2560 pixels. The **application window** width is also correctly set to 2560. However, the **X Screen 0** uses a smaller width, so it cuts off the screen. I can only see the left portion of the RetroArch window. If I set a custom xrandr resolution, on a console, that is 2560 pixels wide **before** starting RetroArch, then the X Screen 0 has the correct width and it works fine. All this, mind you, having only VGA-0 connected as to avoid multi monitor issues.

**Comment by Kopert** [comment]
@alphanu1 thank you for the work.

The crashing problem is fixed, but the incorrect screen size is still an issue, even on a single-monitor setup. The current test results are:

1. Using a multi-monitor setup, crashes are gone. RetroArch no longer crashes on OpenGL. Vulkan is no longer crashing but the screen freezes; I understand that Vulkan is not officially supported and this is not a problem, but if I do the fix described on the problem below, then Vulkan works.
2. I started RetroArch and the image on the TV is messed up: I can see the RetroArch menu on the left side of the screen, sometimes stretched vertically, while about 70% of the screen to the right is just garbled garbage, usually flickering. Loading a game and going through resolution changes flickers the screen as normal, but the issue remains.
 
Regarding this problem, I tried a few different things to diagnose the issue:

- I can fix it by closing RetroArch after this occurs, using xrandr to set a 2560-pixels-wide resolution, then restarting RetroArch.
- For some reason, however, doing this on startup before running RetroArch causes another problem - RetroArch is only shown on the bottom half of the screen. The fix **only** seems to work if you set the 2560 resolution **after** starting RetroArch once and closing it.
- This problem also occurs with a single monitor connected.
- If, after applying the fix successfully I close RetroArch, use xrandr to set the "d_mo" mode (it looks ok) and I start up RetroArch again, the problem returns.

The screenshots below seem show the problem: the first was taken as soon as RetroArch started; the second was taken after a game was loaded. Notice how the screen area ends before the entire RetroArch window is shown.

![whilerunning](https://user-images.githubusercontent.com/26443632/114411118-4922c000-9b82-11eb-8adf-527f82769ba6.png)
![castlevania](https://user-images.githubusercontent.com/26443632/114411130-4c1db080-9b82-11eb-9028-66510b3614d7.png)

**Comment by alphanu1** [comment]
@Kopert This last issue should now be fixed. Just waiting for the PR to be merged.

In the mean time you can test from my reop.  https://github.com/alphanu1/MME4CRT

**Comment by Kopert** [comment]
@alphanu1 thanks again for the effort. I'll do more tests, but I initially ran into some problems.

If I run RA from your repo and I have only a single monitor connected, using OpenGL, I get this when trying to start up:

```
X Error of failed request:  BadMatch (invalid parameter attributes)
  Major opcode of failed request:  140 (RANDR)
  Minor opcode of failed request:  7 (RRSetScreenSize)
  Serial number of failed request:  21
  Current serial number in output stream:  22
```

If I try it with two monitors connected (my usual setup), I get this:
```
X Error of failed request:  BadRRCrtc (invalid Crtc parameter)
  Major opcode of failed request:  140 (RANDR)
  Minor opcode of failed request:  20 (RRGetCrtcInfo)
  Crtc id in failed request: 0x0
  Serial number of failed request:  18
  Current serial number in output stream:  18
```

I will perform more testing and let you know if I discover anything else.

**Comment by alphanu1** [comment]
@Kopert Going down a different route. This is a full implementation of Clamity's Switchers. Using the new Switchres 2 library by him and the group. I had a little help in it 😃 

You need to compile and install the [Switchres](https://github.com/antonioginer/switchres) library. Then try [MME_SR2](https://github.com/alphanu1/MME_SR2 )

Off the bat you will notice better resolutions, faster switching and better compatibility. You can edit the switches.ini to setup monitor profiles too. 

Let me know how you get on?

**Comment by Kopert** [comment]
@alphanu1 First of all, thank you so much for still spending your time working on this. It is a very important feature to me. I was just yesterday trying to fiddle with your code but my C is very rusty these days. I think that your approach to using this external library seems to be a great idea to remove most of the boilerplate system-dependent stuff from your plate. It would probably end up being more compatible overwall, I'd wager.

Just running RetroArch on default with your new code causes a completely new behavior: the main display (the LCD screen that I use) gets turned off and only the CRT is on (this behavior is perfectly fine, it's just new). The image on the CRT, however, is wrong: it extends to the left, so I only get the right side of the image, stretched horizontally. Trying to run a game like this left it as a small square on the top left of the screen, still stretched, with most of it extending outside of the CRT, as if my main screen was still on.

If I unplug the LCD screen before running RetroArch, then it _mostly_ works! Results are as follows:

Good results:
- SNES games look normal;
- PlayStation games look normal, including switches between progressive and interlaced;
- Exiting RetroArch restores a normal, usable resolution on the CRT.

Bad results:
- Game Boy games are no longer a small, pixel-perfect square image in the center and, instead, extend to the entire screen;
- Neo Geo Pocket games send the CRT into an unsupported resolution (reported modeline was 7680x456_60i while my superresolution is set t 2560) when they used to run in progressive scan before;
- Game Boy Advance games are presented in 2560x480i instead of the nice 240p they ran at before, being squished vertically;
- The main RetroArch menu is now interlaced instead of progressive, which, in my personal opinion, is a downgrade, but I can live with it if there's no way around it.

Having to unplug the LCD before using it is a downside to me - because it is cumbersome and because the Linux interface moves to the CRT TV, causing it to flash on screen between resolution switches - but it is still a much better result than before.

I then tried messing with the switchres.ini file as you suggested. Changing "display" to "VGA-0" made both screens work normally and RetroArch to only be presented on the CRT screen, which is perfect! The remaining issues are the ones I mentioned before - such as specific systems' resolutions and RA's main menu.



With all of this said, I'd suggest the following, if I may:
- Fix the resolution issues with the non-standard systems such as GB, GBA and Neo Geo Pocket;
- Make the RetroArch main menu operate at 240p instead of 480i (or give us a choice);
- Allow the user to have the switchres.ini file at the ~/.config/retroarch directory along with all the other RetroArch config files.

Since I can make my non-supported multi-screen setup work just as before just by setting the correct display on the switchres.ini file, and it works fine if you only have one screen connected, I think these changes would be enough to meet everyone's needs.

Please do let me know if you'd like me to do any more testing or to try anything else on my end. Thanks for your time!

**Comment by alphanu1** [comment]
@Kopert Thanks. 

So, we will be adding monitor index support to switchers 2.0. This means you wont have to set these setting in the switchres.ini.

I have removed the low res hand held work around for now. While we are in the early stages of testing at least anyway. I will work them back in once everything else is working as it should.

There is a lot of conversation about the menu resolution. Some what 240p others want 480i and then also people want 480p. So,  I have decided to removed it. The menu will start in whatever you desktop resolution is. This also fixed issues for dual and tri sync monitors.

I have pushed a few changes since you last tested. Please have another go and let me know.

Thanks for doing some of this testing.

Hopefully, the Libretro team will except this when it is time to raise a PR as this now requires an external library. However, this is a much better implementation. More stable and is not limited to just X. The windows / Linux implementation is no longer different. Dynamic modes are now available for windows and so is monitor selection instead of being stuck with primary monitor only.

@twinaphex @hunterk Please could you let me know your thoughts. [New Implimentation](https://github.com/alphanu1/MME_SR2/blob/master/gfx/video_crt_switch.c)

**Comment by hizzlekizzle** [comment]
It's fine by me. Since it's GPLv2+, we could potentially link it in as a vendor-ed dependency.

**Comment by Kopert** [comment]
@alphanu1 
> So, we will be adding monitor index support to switchers 2.0. This means you wont have to set these setting in the switchres.ini.

Sounds great. I look forward to testing it.

> I have removed the low res hand held work around for now. While we are in the early stages of testing at least anyway. I will work them back in once everything else is working as it should.
 
I figured this might've been the case. I'll test that extensively once it's implemented.

> There is a lot of conversation about the menu resolution. Some what 240p others want 480i and then also people want 480p. So, I have decided to removed it. The menu will start in whatever you desktop resolution is. This also fixed issues for dual and tri sync monitors.
> 
> I have pushed a few changes since you last tested. Please have another go and let me know.

So, I was testing this before since I saw you had made a few commits today. Some test results:

- If I have my TV set to 720x480i and I start RetroArch, the main menu only takes half of the screen, both horizontally and vertically, so it is small, centered and not stretched (the on-screen notifications are shown on the correct position, outside of this small interface section, stangely enough);
- If I have my TV set to 320x240p and I start RetroArch, the main menu works as it used to in this resolution;
- If I start a SNES game and return to the main menu, it looks great at 2560x239p; it is strange, but if I set my CRT to 2560x240p before starting RetroArch, the interface is squished horizontally; on-screen notifications are, again, shown "outside" of the squished interface, in the correct location; for some reason, the interface works if coming out of SNES at 2560x239p, but not if I set it manually before starting RetroArch;
- If I start a PlayStation game and quit it while it is at 480i, the resolution stays at 2560x480i and the menu is squished horizontally, the same way as if I started RetroArch while using a superresolution.

I think you _need_ to set a resolution when quitting content because otherwise you're going to keep using the last content's resolution, which might cause issues. If you want to use the original desktop resolution then you'd need to store that at startup so you can set it back whenever running content is closed. This might not be strictly necessary, but I think it would look better if the RetroArch's main menu resolution was consistent even after quitting content.

I understand if you'd rather not mess with the main menu resolution directly and just use whatever is being used by the desktop, as it gives control to the user and makes things easier for your code. I'd be happy with just using 2560x240p as my desktop resolution and have RetroArch use that, but the main menu would have to fill the screen correctly, and RetroArch would need to set this resolution back after closing content, I think.

If these two changes could be made then I'd just use 2560x240p on my desktop and have RetroArch always use it when not running content.

**Comment by alphanu1** [comment]
@Kopert I have added back in the Handheld check. 

Please let me know if this works as well as it used to?

**Comment by Kopert** [comment]
@alphanu1 test results (at 2560 superresolution):

- Game Boy games seem to look fine. Resolution is at 2560x200#59.73 when it used to be at 2560x144@59.73 on 1.9.0, but they look the same as before.
- Game Boy Advance games are running at the same resolution as before, 2560x200#59.73. They look different, but in a good way: they are smaller horizontally to account for the system's aspect ratio. I used to have a per-system shader adjustment to correct the aspect ratio of some handhelds because most of them were stretched horizontally. So it works better now!
- Game Gear games, like Game Boy games, are at a different resolution (2560x200@59.92 instead of 2560x144_59.92) but otherwise look exactly the same as on 1.9.0. They do not have the same aspect ratio correction present on GBA, so I'm still using my shader to adjust the horizontal size, but this is the same behavior as on 1.9.0.
- Neo Geo Pocket is at the exact same resolution as before - 2560x200@60.25 - and has the aspect ratio correction that I found on the GBA games.
- Atari Lynx games did not work before because they used to be at 2560x240@75, which sends my CRT into an unsupported resolution. They're different now - resolution is set at 2560x200@65 - but that refresh rate is still to high for my TV so its unusable. I don't know if you intend to do something on this front or not (it didn't work before anyway) but just letting you know, since you seem to have limited the refresh rate somewhat, since it's lower than before. If you could lower it further to 60hz it'd probably work here as well.
- Wonderswan is a similar deal to Lynx - originally it would send my TV into 75hz which is unsupported, now it sends it to 65hz, which also doesn't work for me.

Overall, looks like great results so far; I'm really pleased with how the whole things is coming together. I don't know if the aspect ratio fix on GBA and Neo Geo Pocket was intentional or not but, if it was, you could consider applying it to Game Gear and Game Boy games as well. But this would be just a nice-to-have.

Other than that, I still have to set VGA-0 on the .ini file and the main menu interface still does not fill the screen, but I assume you didn't work on those yet.

**Comment by alphanu1** [comment]
> Other than that, I still have to set VGA-0 on the .ini file and the main menu interface still does not fill the screen, but I assume you didn't work on those yet.

Not yet.

> * Atari Lynx games did not work before because they used to be at 2560x240@75, which sends my CRT into an unsupported resolution. They're different now - resolution is set at 2560x200@65 - but that refresh rate is still to high for my TV so its unusable. I don't know if you intend to do something on this front or not (it didn't work before anyway) but just letting you know, since you seem to have limited the refresh rate somewhat, since it's lower than before. If you could lower it further to 60hz it'd probably work here as well.
> * Wonderswan is a similar deal to Lynx - originally it would send my TV into 75hz which is unsupported, now it sends it to 65hz, which also doesn't work for me.

So Wonderswan and Linx have not worked in the past?

**Comment by Kopert** [comment]
> So Wonderswan and Linx have not worked in the past?

They haven't, unless I turned off switchres.

**Comment by alphanu1** [comment]
@Kopert At least this is not broken functionality but can be looked into for improvement.

**Comment by Kopert** [comment]
> At least this is not broken functionality but can be looked into for improvement.
Agreed.

**Comment by Kopert** [comment]
@alphanu1 sorry, I found a new problem that wasn't occurring yesterday. When running a PlayStation game, when it switches from 480i to 240p, the image becomes horizontally squished as if my superresolution wasn't taken into consideration. The problem does not happen if I use NATIVE resolution.

**Comment by alphanu1** [comment]
@Kopert The main reason this wont be working is that Switchres has been updated to give RA some more option. Please clone, make install the latest from [switchres](https://github.com/alphanu1/switchres)

Monitor index is controlled form RA but all other options are controlled from the switchres.ini. Which can be add to either /ect/ , ./ini or the working folder (RA folder). The dot clock min controls super resolutions which are not locked dynamically integer scaled. if you want super resolutions set this to 25.0. The monitor profile controls KHz.

[Switchres Monitor Presets](https://gitlab.com/groovyarcade/support/-/wikis/3-Post-Installation-and-Maintenance/3.5-Monitor-Presets)

We are still adding more features which is why this is only Ver 0.1

Let me know how you get on with the above.

**Comment by alphanu1** [comment]
hey @Kopert,

I am sure there will be more development or at least custom monitor presets coming. If you look at the monitor presets page you will see values as follows.

The last four number are the min and max values supported. min/max progressive lines and the min/max interlaced lines. At some point we will get profiles allowing as low as 144. but just not yet. I get the normal 288p resolution when I run GB

`crt_range0 15625-15750, 49.50-65.00, 2.000, 4.700, 8.000, 0.064, 0.192, 1.024, 0, 0, 192, 288, 448, 576`

you can try editing crt_range_0 in your ini so it looks like this.

`crt_range0 15625-15750, 49.50-65.00, 2.000, 4.700, 8.000, 0.064, 0.192, 1.024, 0, 0, 144, 288, 448, 576`

The min dot clock set a min value for it. which mean the width will be calculated on a native width scale. Like my old dynamic version. meaning pixel perfect scaling. 

If you want to force 2560 you need to edit the usermode and add the following value 2560x0 however this is not always going to be an exact integer scale. 

The menu is an issue and will be worked on eventually.

**Comment by Kopert** [comment]
Hey @alphanu1, great job on everything so far. I'm starting to understand your systems better now, I believe.

> The menu is an issue and will be worked on eventually.

Understood. When you look at this, could you also take a look at the resolution the screen is set to when closing content? It seems to be based on the resolution of the content that was just closed, and it'd be better if it was set to the same resolution RA started with. But this is just a nice-to-have if you have the time.

> Are you sure you pulled the latest? I had to `git reset --hard origin/master` followed by` ./configure` and then `make` after I rebased and clean all my git commits.

Yes, I did all of that. Had to do a rebase myself, but I'm sure than I'm using your latest.

Now for testing...

I had set the `dotclock_min` to `25.0` before, like I mentioned, and superresolutions seemed to be working fine. I understand they're not set strictly to 2560 so you can have exact integer scaling.

I created a custom `crt_range0`, based on `arcade_15`, and changed the minimum progressive lines to 144. Game Boy games work ok (the aspect ratio seems to be off). GBA games do work, but, when closing the content, RetroArch got sent to my primary monitor. I had to close and restart RA for it to go back to my TV.

Then I tried changing `user_mode` to `2560x0` as you suggested and RetroArch no longer gets sent to my main monitor when closing GBA content. In my particular use case, forcing the superresolutions to 2560 is better than having exact integer scaling horizontally. I can expand on my case if you'd like but I don't think it's relevant as an edge use case, but having this option to force `2560x0` on `switchres.ini` is awesome for me.

Setting `2560x0` also means I don't have to change the `dotclock_min` anymore since I'd not be using your dynamic scaling for exact integer scaling... it makes sense to me now. Still, for other users, it'd be necessary to look into the issue of RA being sent to another display when closing GBA content. No idea why that happens.

I also took this opportunity to change the max refresh rate from 65hz to 60hz and now Atari Lynx and WonderSwan games also run on my TV without sending it into an unsupported resolution.

I don't know how many of these options you intend to bring to RA's interface, if any, but right now, with proper configuring of the `switchres.ini` file, I can get everything (save for menu issues) working properly. Which, along with everything else, leads me to believe that moving to SR2 is indeed the best solution as you proposed, since it allows users to fix their specific issues manually if necessary.

**Comment by alphanu1** [comment]
@Kopert 
**Monitor preset options 15/31KHz now active. Added new meu option.**

Monitor presets can now be chosen from the RA UI. 15KHz and 31KHz will set
arcade_15 and aracde_31 respectively. New option INI, if this is chosen your
monitor preset will be selected from your switchres.ini file.

You will need to build and reinstall switchres `make && make libswitchres && sudo make install`

**Comment by Kopert** [comment]
@alphanu1 I have tested your new code with no `switchres.ini` file, with a super resolution of 2560 and 15khz set on the RA options. Everything works fine except for GB/GBA and handhelds that operate above 60hz.

Then I copied a new default `switchres.ini` file and changed the custom monitor (based on generic_15) to 144 minimum progressive lines and maximum 60hz (my TV can do from 50hz to 60hz). Then I changed the option on RA to INI. Everything is working. Super resolutions from the UI are indeed working as well, so no manual `user_mode` is necessary anymore.

So, everything is working with your changes and some .ini editing, like before. Less editing than before, which is great.

I'll leave a suggestion below but please, feel free to ignore it. I don't know if it would be useful to other users. Its only purpose would be to avoid using the switchres.ini file at all.

But basically, for me, it'd be interesting if there was a new preset - called `tv_15` or something - that went like this

`crt_range0  15625-15750, 49.50-60.00, 2.000, 4.700, 8.000, 0.064, 0.192, 1.024, 0, 0, 144, 288, 448, 576`

Same as generic_15, but with a minimum progressive lines of 144 and a maximum refresh rate of 60hz. And then have all the monitor presets exposed on the RA UI, so I could pick it directly from there.

**Comment by alphanu1** [comment]
Some NTSC games (most) run at around 59hz -61hz can you try the following and let me know how you get on?

`crt_range0 15625-15750, 49.50-61.00, 2.000, 4.700, 8.000, 0.064, 0.192, 1.024, 0, 0, 144, 288, 448, 576`

**Comment by LarsenDX** [comment]
I have several BVMs and PVMs and enjoy playing on them and I do see the same behavior here. CRTSwitchRes works beautifully on 1.9.0. Broken since, on 1.9.1 and 1.9.2 . RA doesn't switch at all.
Fetched the 1.9.0 binary from my bacula backups and using 1.9.0 for now. I guess nobody is using CRTSwitchRes or libretro would take way more flack for pushing new "stable" releases with the CRTSwitchRes feature completely broken like this.

**Comment by Kopert** [comment]
> Some NTSC games (most) run at around 59hz -61hz can you try the following and let me know how you get on?
> 
> `crt_range0 15625-15750, 49.50-61.00, 2.000, 4.700, 8.000, 0.064, 0.192, 1.024, 0, 0, 144, 288, 448, 576`

@alphanu1 had no problem running handheld games at 61hz. In fact, I kept trying increasing the maximum refresh rate and managed to get it working up to 64hz before my TV started crapping out. I don't know about other TVs out there, so maybe pushing all the way to 64hz might be a stretch.

I'll see if I can pull out another CRT TV I have around to test some more.

**Comment by alphanu1** [comment]
@Kopert fantastic. Let me know how you get on?

**Comment by Kopert** [comment]
> @Kopert fantastic. Let me know how you get on?

@alphanu1 So I managed to plug in my backup CRT, it is an older model that looks like crap. But, much to my surprise, I was able to push it all the way up to 70hz before it gave me any problems.

I only have these two test cases with me, unfortunately... among them, the highest I can go is 64hz with my main TV.

**Comment by LarsenDX** [comment]
@alphanu1 Thanks for breaking this in stable. Stay in your forks with your experimental code.

**Comment by Kopert** [comment]
@alphanu1 I saw your commits and the menu is fixed now. It's better than before, even, because the resolution it reverts to is more "constant" than before.

I have a question, however, just curious. I tested some PAL Amiga content that runs at 288p 50hz but switchres sent it to 576 interlaced. Relevant line is:

`Switchres: normal (720x288@49.920128)->(2560x576@49.920128)` 

I was just wondering, would it be possible to keep it at 288p? I assume my TV would support that, but I'm not sure. Is that something I can change in the .ini file? Like I said, just curious.

**Comment by alphanu1** [comment]
@Kopert fry changing the mas progressive from 288 to 290

**Comment by Kopert** [comment]
> @Kopert fry changing the mas progressive from 288 to 290

@alphanu1 that was actually the first thing I tried, thinking it might be a off-by-one issue. After your suggestion, I tried increasing the value one by one. The resolution stays on interlaced until the max progressive lines is at 298; it reverts back to progressive when the max progressive lines is set to 299.

So setting max progressive lines to 299 fixes the issue in my case.

**Comment by alphanu1** [comment]
@Kopert  That's fine. I used to set 300 as my hard set Hight before interlace trigger.

FYI, latest version the menu should be sorted.

**Comment by Kopert** [comment]
@alphanu1 I'm getting some RA crashes on resolution changes. Even had one going back to the menu. This is the log of when it happens:


```
Inside sr_switch_to_mode(2560x240@49.920128)
Switchres: Calculating best video mode for 2560x240@49.920128 orientation: normal

Switchres: [1024]x[ 768]_[60=60.003722Hz] - locked

Switchres: [ 800]x[ 600]_[60=60.315287Hz] - locked

Switchres: [ 800]x[ 600]_[56=56.249600Hz] - locked

Switchres: [ 848]x[ 480]_[60=60.000000Hz] - locked

Switchres: [ 640]x[ 480]_[59=59.939048Hz] - locked

Switchres: [2560]x[ 240]_[60=60.000000Hz]

Switchres: [ 320]x[ 240]_[60=60.000000Hz] - locked

Switchres: [2560]x[ 480]_[60=60.000000Hz]

Switchres: [ 640]x[ 480]_[60=60.000000Hz] - locked

Switchres: [2560]x[ 240]_[60=60.095785Hz]

Switchres: [ 700]x[ 480]_[59=59.939962Hz] - locked

Switchres: [2560]x[ 224]_[59=59.919540Hz]

Switchres: [2560]x[ 144]_[59=59.919540Hz]

Switchres: [2560]x[ 240]_[59=59.919540Hz]

Switchres: (   0)x(   0)_(0=0.000000Hz)
[INFO] Switchres: could not find a video mode that meets your specs
[INFO] sr_switch_to_mode: switching not required
[INFO] [CRT]: Setting Video Screen Size to: 0x0 
[INFO] [CRT]: Setting Aspect Ratio: -nan 
[INFO] [CRT]: SR scaled  X:0 Y:0 
Floating point exception (core dumped)
```

**Comment by alphanu1** [comment]
@Kopert should be fixed now

**Comment by Kopert** [comment]
@alphanu1 I'm sorry, but it is still crashing for me if I keep the maximum progressive lines to 300.

I did, on both `switchres` and `MME_SR2` repos, a full set of `git pull`, `git reset --hard`, `make clean,` `configure`, `make`, `make install` (for `switchres` first) to make sure it's all up to date.

I'm using a custom monitor range with the `INI` option with these settings:

`15625-15750, 49.50-64.00, 2.000, 4.700, 8.000, 0.064, 0.192, 1.024, 0, 0, 144, 300, 448, 576`

And I'm starting RA with a resolution of 2560x240 60hz, with timings provided by switchres.

RetroArch starts fine, the menu is perfect. However, it crashes when trying to start any SNES, SMS or GBA game (I haven't tested any others). It **does not** crash if I start some PAL Amiga game that runs at 720x288@49.92.

If I switch it back to 288 max progressive lines it no longer crashes, but the PAL Amiga game reverts to interlaced.

**Comment by Kopert** [comment]
@alphanu1 I built everything anew after your 0.2 rebase. Here's a list of bugs I was able to find:

- PAL Amiga game, with original resolution of 720x288_49.92hz, when running on a super resolution of 1920 or 2560, gets sent to 576 interlaced instead of staying on 288p; does not happen if resolution is set to dynamic, native or 3840 (stays at 720x288p in the first two cases, goes to 3840x288p on the latter, as expected)
- When using a super resolution of 1920, after closing content, the RetroArch window gets sent to my HDMI-A-0 display instead of staying on my VGA-0 display; does not happen on any other super resolution (native, dynamic, 2560, 3840)
- Closing content from a core and running content from the same core (or from another core that uses the same resolution as the first one) causes the resolution to not switch on the second run, staying on the RA menu resolution; does not happen if the second content being run is on a different core that uses a different resolution; as long as each subsequent content run is of a different resolution than the previous one, it works
- Handheld cores are not being shown in the correct aspect ratio, filling the screen horizontally (tested with GB, GBA and Game Gear; console **does** show the message "Setting Aspect Ratio: 1.111111 " for GB games, but they're still stretched horizontally

I'll let you know if I find anything else.

**Comment by Kopert** [comment]
@alphanu1 I saw the 0.5 release. Did a full clean, rebase etc.; just to be sure, I removed the installed switchres library from my computer since it is now supposed to be baked into RA. I then proceeded to testing. Here are the testing steps I took and the results:

- Started RA, made sure it was set to `3840` super resolution and `INI` settings (this is what I had the best results with before), made sure `Integer Scale` was set to `ON` and `Aspect Ratio` was set to `Core provided`, then restarted RA
- Ran some PlayStation content; resolution changes happened as expected, closed the content, no issues
- Started some SNES content, resolution went to 239p as expected; tried to close the content and got a **crash**:

```
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 27 seconds.
[INFO] Saving runtime log file: /redacted.lrtl
[INFO] [CORE]: No content, starting dummy core.
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 00 seconds.
[INFO] [Core]: Unloading game..
[INFO] [Core]: Unloading core..
[INFO] [Core]: Unloading core symbols..
[INFO] [Core Options]: Saved core options file to "/redacted/retroarch-core-options.cfg"
Segmentation fault (core dumped)
```

- Restarted RA, resolution was at 3840x239p as the previous content left it when it crashed, the menu was displayed normally
- Tried to start some PAL Amiga content, resolution was correctly set to 3840x288_49.92hz, content looks fine
- Closed the previous content and tried to start some GB content using Gambatte. The resolution **did not change** and the content filled the screen. Logs showed as follows:

```
[INFO] [CRT]: Requested Reolution: 160x144@59.727501 
[INFO] [CRT]: SR init 
[INFO] [CRT]: CRT Mode: 4 - Seleted from ini 
[INFO] [CRT]: SR init_disp 
[INFO] SRobj: RA Monitor Index: 1
[INFO] [CRT]: SR Disp Monitor Index: 1  
[INFO] [CRT]: SR rtn 1 
[INFO] Switchres: Modeline "3840x144_59 15.648605KHz 59.727501Hz" 78.039594 3840 3996 4363 4987 144 194 197 262   -hsync -vsync
[ERROR] XRANDR: <2> (add_mode) [WARNING] this screen is managed by <1310731552>
[ERROR] sr_refresh_display: error refreshing display
[INFO] [CRT]: SR failed to switch mode[INFO] [CRT]: Setting Video Screen Size to: 92160x144 
[INFO] [CRT]: Setting Aspect Ratio: 640.000000 
[INFO] [CRT]: SR scaled  X:24 Y:1 
[INFO] [Video]: Setting refresh rate to: 59.728 Hz.
```

- Closed the content, tried running it again, got the same **crash** as previously
- Restarted RA, went straight into trying to run the same GB content as before. The image is heavily squished horizontally (this includes the RA menu if I press F1) and I can barely see anything. Logs show:

```
[INFO] [CRT]: Requested Reolution: 160x144@59.727501 
[INFO] [CRT]: SR init 
[INFO] [CRT]: CRT Mode: 4 - Seleted from ini 
[INFO] [CRT]: SR init_disp 
[INFO] SRobj: RA Monitor Index: 1
[INFO] [CRT]: SR Disp Monitor Index: 1  
[INFO] [CRT]: SR rtn 1 
[INFO] Switchres: Modeline "3840x240_59 15.648605KHz 59.727501Hz" 78.039594 3840 3996 4363 4987 240 242 245 262   -hsync -vsync
[INFO] sr_refresh_display: mode was added
[INFO] sr_switch_to_mode: successfully switched to 3840x240@59.727501
[INFO] [CRT]: Setting Video Screen Size to: 46080x240
[INFO] [CRT]: Setting Aspect Ratio: 192.000000 
[INFO] [CRT]: SR scaled  X:12 Y:1 
[INFO] [CRT]: Setting Video Screen Size to: 160x144 
[INFO] [CRT]: Setting Aspect Ratio: 1.111111 
[INFO] [Video]: Setting refresh rate to: 59.728 Hz.
```

- Restarted RA and tried to run GBA content using mGBA. Also got a heavily horizontally squished image (including the RA menu), same as GBA.
- Restarted RA and tried to run Game Gear content using Genesis Plus GX. The aspect ratio was wrong (it was strecthed horizontally) but otherwise worked fine. I got the same results with WonderSwan and Atari Lynx. This result is fine for me as I have other ways of fixing the aspect ratio if it is _wider_ than it is supposed to be.
- Restarted RA. Ran some SNES content, closed it, ran the same content again, closed it, ran some PlayStation content, closed it, got a **crash**. Logs at the end show:

```
[INFO] [GL]: VSync => OFF
[INFO] [GLX]: glXSwapInterval(0)
[INFO] [GL]: VSync => ON
[INFO] [GLX]: glXSwapInterval(1)
[INFO] [PulseAudio]: Pausing.
[ERROR] XRANDR: <2> (set_timing) [ERROR] width is above allowed maximum (167186432 > 16384)
[ERROR] XRANDR: <2> (set_timing) [ERROR] height is above allowed maximum (22190 > 16384)
free(): invalid pointer
Aborted (core dumped)
```

- Restarted RA, ran some PlayStation content, closed it, ran it again, closed it. Now RA was mirrored in my main display alongside my VGA-0 output. I tried to close RA and got another crash.

I _did not_ test any other options for super resolution, so I don't know if the previous bugs are still present (since most of them did not happen on 3840). I can test those other resolutions again at a later point.

I hope this testing helps. Let me know if I can help in any other way.

**Comment by Kopert** [comment]
The previsouly reported issues on 0.5 are still present on 0.6

**Comment by alphanu1** [comment]
@Kopert some new checks have been put in place to GB and GBA content. Try these using arcade_15 from the RA menu. Also remember you can save core overrides when you want o use INI over an monitor preset. Or different presets per core. More UI presets will be added down the road and ini over rides per core if one should exist systemname_switchres.ini

thanks for testing and reporting though it does help.

**Comment by Kopert** [comment]
@alphanu1 I just want to help in any way I can.

I don't think my issues are related to monitor presets. I changed it back to 15khz in the RA menu and the problems persisted.

The crashes and resolutions failing to change only happen if I load different types of content in succession, for example, if I load a SNES game, then close it and load a GBA game... the resolution doesn't change on the second content and if I keep loading different contents it crashes. If I restart RA every time before I start a game, these issues don't happen.

The problem with GB/GBA games being squished horizontally only happens if I'm using a super resolution that is not NATIVE or DYNAMIC. As you can see in the logs, switchres is trying to set the wrong resolution for the game (after successfully changing the display resolution), as in this message: `Setting Video Screen Size to: 46080x240`

And I'm concerned with your solution for handhelds being based on the name of the core, because, for example, Game Gear games use the same cores used for Master System and Genesis, but that is an issue for another time.

**Comment by alphanu1** [comment]
HH workaround is experimental and only for GB and GBA ATM. Can you test a lower super width? say 2560.

This Setting Video Screen Size to: 46080x240 is a very odd situation. It sounds a though 3840 is not hard set and the dot clock min is multiplying it.

Can you send me you current ini settings?

**Comment by Kopert** [comment]
@alphanu1 the `dotclock_min` of my .ini file was at `0`. You can find my `switchres.ini` file at https://pastebin.com/qYtAN7vV

For further testing, I **removed my switchres.ini file completely** and made sure it was set to 15khz on the RA menu. Then I did the following tests, restarting RA before each test:

- NATIVE: GBA content runs fine with the correct aspect ratio; screen is set to 320x240
```
[INFO] sr_switch_to_mode: successfully switched to 320x240@59.727501
[INFO] [CRT]: Setting Video Screen Size to: 320x240 
[INFO] [CRT]: Setting Aspect Ratio: 1.333333 
[INFO] [CRT]: SR scaled  X:1 Y:1 
[INFO] [CRT]: Setting Video Screen Size to: 240x160 
[INFO] [CRT]: Setting Aspect Ratio: 1.500000
```

- DYNAMIC: GBA content runs fine with the correct aspect ratio; screen is set to 320x240
```
[INFO] sr_switch_to_mode: successfully switched to 320x240@59.727501
[INFO] [CRT]: Setting Video Screen Size to: 320x240 
[INFO] [CRT]: Setting Aspect Ratio: 1.333333 
[INFO] [CRT]: SR scaled  X:1 Y:1 
[INFO] [CRT]: Setting Video Screen Size to: 240x160 
[INFO] [CRT]: Setting Aspect Ratio: 1.500000
```

- 1920: GBA content runs squished horizontally; screen is set to 1920x240
```
[INFO] sr_switch_to_mode: successfully switched to 1920x240@59.727501
[INFO] [CRT]: Setting Video Screen Size to: 11520x240 
[INFO] [CRT]: Setting Aspect Ratio: 48.000000 
[INFO] [CRT]: SR scaled  X:6 Y:1 
[INFO] [CRT]: Setting Video Screen Size to: 240x160 
[INFO] [CRT]: Setting Aspect Ratio: 1.500000
```

- 2560: GBA content runs squished horizontally (more squished than at 1920); screen is set to 2560x240
```
[INFO] sr_switch_to_mode: successfully switched to 2560x240@59.727501
[INFO] [CRT]: Setting Video Screen Size to: 20480x240 
[INFO] [CRT]: Setting Aspect Ratio: 85.333336 
[INFO] [CRT]: SR scaled  X:8 Y:1 
[INFO] [CRT]: Setting Video Screen Size to: 240x160 
[INFO] [CRT]: Setting Aspect Ratio: 1.500000
```

- 3840: GBA content runs squished horizontally (more squished than at 2560); screen is set to 3840x240
```
[INFO] sr_switch_to_mode: successfully switched to 3840x240@59.727501
[INFO] [CRT]: Setting Video Screen Size to: 46080x240 
[INFO] [CRT]: Setting Aspect Ratio: 192.000000 
[INFO] [CRT]: SR scaled  X:12 Y:1 
[INFO] [CRT]: Setting Video Screen Size to: 240x160 
[INFO] [CRT]: Setting Aspect Ratio: 1.500000
```

**Comment by alphanu1** [comment]
@Kopert just pushed a fix that should fix the weird video resolution/aspect bug. Please can you confirm this is now fixed for you?

**Comment by Kopert** [comment]
#@alphanu1 Testing with commit 4c390ea, the problem with different contents not switching resolutions and RA crashing seems to be fixed, thanks! I could not reproduce these problems anymore.

Now trying to provide you with more data regarding the GB/GBA issue:

The GB/GBA issue is still present and apparently if I start one of those, close it, then try to start SNES content (without restarting RA), the SNES content ends up squished as well, as if the Aspect Ratio correction from GBA is still active.

The reverse is also true: if I restart RA, run SNES content (it looks fine), close it, then start GBA content, it does not squish horizontally, as the Aspect Ratio correction remains disabled.

Just a suggestion: I know the current handheld handling is an experimental workaround and might end up working completely different, but might I suggest adding an option to the RA UI for SwitchRes, or to the INI file, to disable the Aspect Ratio correction for handheld content? In case some people need/prefer the unaltered resolution.

**Comment by alphanu1** [comment]
@Kopert I have now removed the HH code as the GA team also thing this will be better handles via the switchres.ini. Hopefully, we can get more options in down the road.

If you can test and let me know if everything is working OK. 

And if you don't mind commenting on you testing an latest results over at the [PR](https://github.com/libretro/RetroArch/pull/12403) that would be great.

**Comment by AlexFolland** [comment]
The pull request has since been merged.  Is this working now, leading this ticket to be ready to be closed?

**Comment by alphanu1** [comment]
Yes this ticket should be closed now.

**Comment by alexb3d** [comment]
I still have the same problems that I told you before.
It works, although on a monitor presents some errors. I have not been able to use it on a TV.

1. You have to use the advanced menu, activate the option in Basic Menu does not do anything.

2. You have to activate the scaling video option, because if not, [the image is crushed.](https://ibb.co/WDWywd5)

3. When closed, return to the resolution of the desktop correctly. But when it restarts and then closes, it stays at 240

4. When the game is loaded does not activate 240P, you have to close the game and reload it.

5. Activating the menu works well without loading a game, when the game and the menu is activated, [you can not distinguish.](https://ibb.co/7yLCyJR)

**Comment by alphanu1** [comment]
@alexb3d My assumption here is that you are using windows!? This all leads to the point that you resolution either not setup correctly. Things have changed over to Calamity's Switchres. So resolutions must be installed as dynamic, as you would for GroovyMame. You must have a switchres.ini in you RetroArch working folder and setup with the presets. If you want to you a monitor preset from the ini, you must choose INI as the CRTSwitchres menu. otherwise, choose a preset. Once this is done you also need to select you monitor index, especially if you have more then one monitor connected. 

[CRTEmudriver Documentation ](https://geedorah.com/eiusdemmodi/forum/index.php) Make sure you install you resolution as outlined.

[New SR2 CRTSwithres Documentaion](https://docs.libretro.com/guides/crtswitchres/)

I still believe there is a but when in a multiple GPU environment.

**Comment by alexb3d** [comment]
[I keep using Linux. ](https://github.com/libretro/RetroArch/issues/12244#issuecomment-816400471)

In version 1.9.8 continues to pass the same error with the integrated Intel.

Tested with a GeForce GT620 and a Radeon HD 5500. It does not serve, it does not serve the CRT, the program closes or directly the monitor is turned off without the possibility of recovering.

**Comment by alexb3d** [comment]
I just tried the stable version, 1.9.8 of the PPA and the AppImage.
With the GPU the integrated, a GeForce and an ATI.
In the 3 it gives me the error:

`[Video]: Setting refresh rate to: 0.000 Hz.`

I do not want to be a nuisance and I do not know where I can help. If you want I give you remote access to my PC.



log:
```
[INFO] [Config]: Loading default config.
[INFO] [Config]: Looking for config in: "/home/lex/.config/retroarch/retroarch.cfg".
[INFO] RetroArch 1.9.8 (Git fa5ec2d4b1)
[INFO] === Build =======================================
[INFO] CPU Model Name: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz
[INFO] Funcionalidades:  MMX MMXEXT SSE SSE2 SSE3 SSSE3 SSE4 SSE4.2 AES AVX
[INFO] Built: Aug 22 2021
[INFO] Version: 1.9.8
[INFO] Git: fa5ec2d4b1
[INFO] =================================================
[INFO] [Input]: Found input driver: "x".
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] Versión de la API libretro: 1
[INFO] API usada en la compilación: 1
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Video @ fullscreen
[ERROR] [Wayland]: Failed to connect to Wayland server.
[INFO] [GLX]: GLX_EXT_swap_control_tear supported.
[INFO] [GL]: Found GL context: x
[INFO] [GL]: Detecting screen resolution 3200x1080.
[INFO] [GLX]: Window manager is GNOME Shell.
[INFO] [XINERAMA]: Xinerama version: 1.1.
[INFO] [XINERAMA]: Xinerama screens: 2.
[INFO] [GLX]: Using Xinerama on screen #0.
[INFO] [GLX]: X = 0, Y = 0, W = 1280, H = 960.
[INFO] [GLX]: Using windowed fullscreen.
[INFO] [GLX]: Found swap function: glXSwapIntervalEXT.
[INFO] [GL]: Vendor: NVIDIA Corporation, Renderer: GeForce GT 620/PCIe/SSE2.
[INFO] [GL]: Version: 4.6.0 NVIDIA 390.144.
[INFO] [GL]: Using resolution 1280x960
[INFO] [GL]: Default shader backend found: glsl.
[INFO] [Shader driver]: Using GLSL shader backend.
[INFO] [GLSL]: Checking GLSL shader support ...
[WARN] [GL]: Stock GLSL shaders will be used.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GL]: Using 4 textures.
[INFO] [GL]: Loaded 1 program(s).
[INFO] [GL]: Using GL_RGB565 for texture uploads.
[INFO] [udev]: Pad #0 (/dev/input/event8) supports force feedback.
[INFO] [udev]: Pad #0 (/dev/input/event8) supports 16 force feedback effects.
[INFO] [Joypad]: Found joypad driver: "udev".
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [DBus]: Suspended screensaver via DBus.
[INFO] [Video]: Found display server: x11
[INFO] [PulseAudio]: Requested 24576 bytes buffer, got 18432.
[INFO] [Display]: Found display driver: "gl".
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [Font]: Using font rendering backend: freetype.
[INFO] [SRAM]: No se guardará la SRAM.
[INFO] [Playlist]: Cargando historial: [/home/lex/.config/retroarch/content_history.lpl].
[INFO] [Playlist]: Cargando historial: [/home/lex/.config/retroarch/content_music_history.lpl].
[INFO] [Playlist]: Cargando historial: [/home/lex/.config/retroarch/content_video_history.lpl].
[INFO] [Playlist]: Cargando historial: [/home/lex/.config/retroarch/content_image_history.lpl].
[INFO] [Playlist]: Cargando favoritos: [/home/lex/.config/retroarch/content_favorites.lpl].
[INFO] [PulseAudio]: Pausing.
[INFO] [GLX]: Resized fullscreen resolution to 1280x960.
[INFO] [CRT]: Requested Reolution: 320x240@60.000000 
[INFO] [CRT]: SR init 
[INFO] [CRT]: CRT Mode: 3 - pc_31_120 
[INFO] [CRT]: SR init_disp 
[INFO] [CRT]: RA Monitor Index Auto: auto
[ERROR] XRANDR: <1> (xrandr_timing) [ERROR] missing X11_LIBRARY library
[INFO] [CRT]: SR Disp Monitor Index Auto: Auto  
[INFO] [CRT]: SR rtn 0 
[INFO] [CRT]: SR failed to init 
[INFO] [CRT]: Setting Aspect Ratio: 1.333333 
[INFO] [CRT]: Setting Video Screen Size to: 320x240 
[INFO] [Video]: Setting refresh rate to: 0.000 Hz.
[INFO] [GLX]: Resized fullscreen resolution to 1280x960.
[INFO] [PulseAudio]: Unpausing.
[INFO] [PulseAudio]: Pausing.
[INFO] [PulseAudio]: Unpausing.
[INFO] [Config]: Se ha guardado una nueva configuración en "/home/lex/.config/retroarch/retroarch.cfg".
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 00 seconds.
[INFO] [PulseAudio]: Pausing.
[INFO] [Core]: Unloading core..
[INFO] [Core]: Unloading core symbols..
[INFO] [XINERAMA]: Xinerama version: 1.1.
[INFO] [XINERAMA]: Xinerama screens: 2.
[INFO] [XINERAMA]: Saved monitor #0.
[INFO] [Video]: Does not have enough samples for monitor refresh rate estimation. Requires to run for at least 4096 frames.
[INFO] [Video]: Does not have enough samples for monitor refresh rate estimation. Requires to run for at least 4096 frames.
```

**Comment by alphanu1** [comment]
It shows in the Log that you are missing X libraries and SR failed to init

**Comment by alexb3d** [comment]
It still does not work. 1.9.9
![Captura de pantalla de 2021-09-08 02-14-37](https://user-images.githubusercontent.com/14365508/132456340-27f98fc3-6d37-4851-be90-551966df29a0.png)

**Comment by alphanu1** [comment]
This is working fine for many other users. It’s all about your OS and setup. 

You need to ensure all X11 libraries are installed. You can get more information over on the GrooveyArcade forum

**Comment by alexb3d** [comment]
The libraries are in the repository. When retroarch is installed does not install the libraries as dependencies. Nobody who use Ubuntu will be able to use this feature if it does not manually install the libraries.

[https://askubuntu.com/questions/1344364/i-am-getting-error-about-x11-library](https://askubuntu.com/questions/1344364/i-am-getting-error-about-x11-library)

`sudo apt-get install libx11-dev libxrandr-dev`

**They are in the repository, they must be installed by default.**

I already have several versions asking for help and could not you tell me this?
Lack of consideration friend.

**Comment by alphanu1** [comment]
No need to be like that. 

Multiple post and responses from myself mention missing X libraries. Without knowing what you do or don’t have installed does not help. Anyway, I am glad you have resolved it.

**Comment by hizzlekizzle** [comment]
it looks like those libs are not currently installed as runtime dependencies for the PPA packages, but I can change that.

**Comment by alphanu1** [comment]
The CRTSwitchRes docs should be updated soon with the new information. Docs [PR](https://github.com/libretro/docs/pull/664)

**Comment by alexb3d** [comment]
@alphanu1  Please do not get me wrong. I'm going to tell you what I have in my head.

Ubuntu and its derivatives is the most used distro, by far.
When an ordinary user use the CRT and does not work, he is not going to look for all this information. "Generate the log, open reports, search for information on the network that is X11, etc."
What he is going to do is say "this does not serve, Retroarch is a shit." That is, bad publicity.

**RetroArch is wonderful and the CRTSwicht is simply amazing**, it has to work well.

_Thank you for all the work you are doing and for your patience._

**Comment by evelyndooley** [comment]
These libraries are also missing in the Flatpak version. I am running on Fedora and even with the libraries installed through DNF, I am getting this line in my log file. 

```[ERROR] XRANDR: <4> (xrandr_timing) [ERROR] missing X11_LIBRARY library```

Attempting to compile from source on fedora was a nightmare, the resulting build was missing input drivers that worked with any controllers and had some other issues. Is there a way to add these libraries to the flatpak version or am i going to have to switch to ubuntu?

**Comment by alexb3d** [comment]
@evelyndooley [Long, time ago is reported ](https://github.com/flathub/org.libretro.RetroArch/issues/183)but, it is Flatpak :/

AppImage is a great option, install the libraries "libx11-dev libxrandr-dev" manually from the Fedora repository.

**Comment by MiltosKoutsokeras** [comment]
Hi all,

I have been bitten also by the error message:

`[ERROR] XRANDR: <1> (xrandr_timing) [ERROR] missing X11_LIBRARY library`

in Debian 11. I build from sources and the deployment machine is different than the build machine (both Debian 11). The reason why this happens, is the the code is explicitly searching for the X11 library name `libX11.so` in file `deps/switchres/custom_video_xrandr.cpp` lines 127, 325:

```C++
m_x11_handle = dlopen("libX11.so", RTLD_NOW);
```

This cannot work on systems that do not have installed the development library and only have the runtime library. The runtime library does not install the symbolic link `/usr/lib/x86_64-linux-gnu/libX11.so`, so if the machine was not used to build the program with all installed dev dependencies, it will fail to run CRT Switch Res, as the library is opened at runtime with `dlopen` instead of being linked on build with the same name. Check the linked libraries with this command:

```Bash
ldd <path/to/your/retroarch/executable>
```

On my system and configuration it prints the following:

```
	linux-vdso.so.1 (0x00007ffd1bbe6000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f6c63892000)
	libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x00007f6c6374f000)
	libXext.so.6 => /lib/x86_64-linux-gnu/libXext.so.6 (0x00007f6c6373a000)
	libXxf86vm.so.1 => /lib/x86_64-linux-gnu/libXxf86vm.so.1 (0x00007f6c63534000)
	libXrandr.so.2 => /lib/x86_64-linux-gnu/libXrandr.so.2 (0x00007f6c63329000)
	libGL.so.1 => /lib/x86_64-linux-gnu/libGL.so.1 (0x00007f6c632a2000)
	libEGL.so.1 => /lib/x86_64-linux-gnu/libEGL.so.1 (0x00007f6c6328b000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f6c630be000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f6c630b8000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6c62f74000)
	libmvec.so.1 => /lib/x86_64-linux-gnu/libmvec.so.1 (0x00007f6c62f48000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f6c62f2e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6c62d67000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f6c64455000)
	libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f6c62d3c000)
	libXrender.so.1 => /lib/x86_64-linux-gnu/libXrender.so.1 (0x00007f6c62b32000)
	libGLdispatch.so.0 => /lib/x86_64-linux-gnu/libGLdispatch.so.0 (0x00007f6c62a7a000)
	libGLX.so.0 => /lib/x86_64-linux-gnu/libGLX.so.0 (0x00007f6c62a46000)
	libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007f6c62a3f000)
	libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f6c62839000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007f6c62822000)
	libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007f6c62815000)
```

As you can see it is linked against `libX11.so.6` which is found, but on code it tries to load `libX11.so` which is missing on systems that did not build the program themselves.

One workaround is to simply install the dev library `sudo apt install libx11-dev`. But I think for consistency with modern Linux configuration the code of switchres needs to change and load without `dlopen` since it is a hard dependency on the functionality and not a plugin.

**Comment by alexb3d** [comment]
@MiltosKoutsokeras Please read the previous message.

**Comment by MiltosKoutsokeras** [comment]
Yes the installation of the development libraries solved the issue. But you should not expect the runtime machine to have the development libraries when it is using a program. Most Linux distributions have 2 different packages for runtime and development files. If you build in a single machine and copy the built artifacts to multiple, you will encounter the issue.

**Comment by alexb3d** [comment]
I do not understand anything, you're talking to me in Chinese. :D 

He who can answer your doubts is @hunterk

**Comment by LibretroAdmin** [comment]
@alphanu1 can this issue be closed by now?

**Comment by alphanu1** [comment]
Yep this can be closed off. @LibretroAdmin

**Comment by gouchi** [comment]
@LibretroAdmin Please close this issue. Thank you.
